nexus/src/external_api/device_auth.rs

```diff
@@ -0,0 +1,262 @@
+/// a browser). If the user is not logged in, we redirect them to
+/// the login page and use the `state` parameter to get them back
+/// here on completion. If they are logged in, serve up the console
+/// verification page so they can verify the user code.
+#[endpoint {
+    method = GET,
+    path = "/device/verify",
+    tags = ["hidden"], // "token"
+}]
+pub async fn device_auth_verify(
+    rqctx: Arc<RequestContext<Arc<ServerContext>>>,
+    params: Query<DeviceAuthVerifyParams>,
+) -> Result<Response<Body>, HttpError> {
+    // If the user is authenticated, serve the console verification page.
+    if let Ok(opctx) = OpContext::for_external_api(&rqctx).await {
+        if opctx.authn.actor().is_some() {
+            return serve_console_index(rqctx.context()).await;
+        }
+    }
+
+    // Otherwise, redirect for authentication.
+    let params = params.into_inner();
+    let state_params = serde_urlencoded::to_string(serde_json::json!({
+        "user_code": params.user_code
+    }))
+    .map_err(|e| HttpError::for_internal_error(e.to_string()))?;
+    let state = Some(format!("/device/verify?{}", state_params));
+    Ok(Response::builder()
+        .status(StatusCode::FOUND)
+        .header(http::header::LOCATION, get_login_url(state))
+        .body("".into())?)
+}
+
+/// Confirm an OAuth 2.0 Device Authorization Grant
+///
+/// This endpoint is designed to be accessed by the user agent (browser),
+/// not the client requesting the token. So we do not actually return the
+/// token here; it will be returned in response to the poll on `/device/token`.
+#[endpoint {
+    method = POST,
+    path = "/device/confirm",
+    tags = ["hidden"], // "token"
+}]
+pub async fn device_auth_confirm(
+    rqctx: Arc<RequestContext<Arc<ServerContext>>>,
+    params: TypedBody<DeviceAuthVerifyParams>,
+) -> Result<HttpResponseOk<()>, HttpError> {
+    let apictx = rqctx.context();
+    let nexus = &apictx.nexus;
+    let params = params.into_inner();
+    let handler = async {
+        let opctx = OpContext::for_external_api(&rqctx).await?;
+        let &actor = opctx.authn.actor_required()?;
+        let _token = nexus
+            .device_auth_verify(&opctx, params.user_code, actor.actor_id())
+            .await?;
+        Ok(HttpResponseOk(()))
+    };
+    apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+pub struct DeviceAccessTokenRequestParams {
+    pub grant_type: String,
+    pub device_code: String,
+    pub client_id: Uuid,
+}
+
+#[derive(Debug)]
+pub enum DeviceAccessTokenResponse {
+    Granted(DeviceAccessToken),
+    Pending,
+    Denied,
+}
+
+/// Request a device access token
+///
+/// This endpoint should be polled by the client until the user code
+/// is verified and the grant is confirmed.
+#[endpoint {
+    method = POST,
+    path = "/device/token",
+    content_type = "application/x-www-form-urlencoded",
```

Comment:
I don't have strong opinions on whether these should appear in the spec or not; we can set `unpublished = true` if you think that's a good idea. Personally I think they should appear, but don't necessarily need full body descriptions if that's not easy. The TS generator, for instance, just skips non-`application/json` endpoints, which I think is reasonable, since they'll be used only by OAuth client libraries, not directly by our client.

The reason for not using `application/json` is that the OAuth protocol requires `application/x-www-form-urlencoded` request bodies, and I would like us not to reinvent that particular wheel. OAuth clients exist for every popular language & framework, and the idea is that we can use off-the-shelf, well-tested client implementations instead of rolling our own _n_ times. That is, we _could_ implement and document our own not-quite-compatible protocol, but I don't see any concrete advantage to doing so. OAuth is not a great protocol, but it's been widely analyzed and implemented, and I think if we're not going to use it, we should have good reasons why we think our own version would be better, more secure, etc. This PR implements only the Device Authorization Grant flow, but we can layer more things on it, such as short-lived access tokens with long-lived refresh tokens. RFD 275 attempts to provide some of this rationale, and is probably the right place for a serious discussion of this issue.

Finally, I suppose it's fair to ask why we should write our own OAuth server but not the client libraries. To this I would respond that (1) we only need one server implementation, whereas for the clients we'd need _n_ clients with specialized token, error, retry, and timeout handling, and (2) to the best of my understanding, we want there to be a single source of truth server-side (CockroachDB) and a single piece of software (Nexus) that handles all client requests and talks to that database, provides authz for it, etc. I did look into off-the-shelf OAuth servers, but none of them seemed suitable for our needs. OTOH, the client libraries are readily available and easy to integrate.